### ISSUE TITLE
Feature request: KiNET V2

### ISSUE THREAD

**Issue by robbavel** [Breina/ha-artnet-led#43]
Would it be possible to support KiNET V2?

In V2 you can also specify a physical output port for the KiNET device (there are 16 port devices that will only output light on port 1 when using KiNET V1

**Comment by Breina** [comment]
Sure, can you give me the spec?

**Comment by robbavel** [comment]
Do you have an idea where i should look for the specs?

I've only used it before on official kinet soft and hardware before

**Comment by Breina** [comment]
With spec I mean a document explaining the protocol in detail so that I can implement it. I can't find one either.

The biggest hurdle in implementing KiNet is that we have no idea how it works.

@jjhuff do you happen to have the KiNet spec, or anything that could help with this?

**Comment by robbavel** [comment]
https://wiki.openlighting.org/index.php/KiNET

This is what i was able to find

Honestly i have no idea if this would he what you need. Or if it would be
V1 or V2

**Comment by jjhuff** [comment]
It seems like KiNet V2 also goes by 'portout'. OLP has implemented it: https://github.com/OpenLightingProject/ola/issues/1544

I took a quick scan and it looks like the protocol is similar on the wire, but I'm not sure of the other differences and how compatible they'd be with this project.

**Comment by Breina** [comment]
>TLDR:
>
>* You can send data as a single universe via multicast.
>* You can send data as a single numbered universe as broadcast (like sACN).
>* You can send data unicast to specific physical output ports.
>* You can send the same data unicast to all physical output ports on a device at once.

So 'portout' is a specific packet of KiNet, not a separate version of the protocol itself. Still V1 though.

![image](https://github.com/Breina/ha-artnet-led/assets/3009914/d9ba424c-8ecf-4316-8f79-21a31ff32175)

Looks like KiNet has a polling and poll-reply system, just like ArtNet is. I can use the OLA implementation to expand the current V1 implementation.

I'll send an email colorkinetics if they don't want to share the protocol with me directly. Worth a shot. :)

**Comment by robbavel** [comment]
I'm going to guess that contacting CK will lead to a dead end.


If its possible to reverse engineer KiNET V2 via wireshark or something. I
have loads of KiNET V1 and V2 equipment available to sniff traffic on (i
would need a lot of help with the process of capturing it tho)

**Comment by Breina** [comment]
Yeah I don't think I succeeded in explaining the sales lady what a protocol is. :)

# A quick protocol reverse engineer guide

1. Start up Wireshark on the same device as QuickPlay Pro
2. Start a new capture on the network interface that is connected to your network (if you have multiple, choose the one that shows most traffic after selecting it)
3. Apply this filter: `udp.port == 6038 || tcp.port == 6038`, this filters all KiNet traffic. This assumes KiNet V2 uses the same port. Click the _Apply_ button.
4. Perform some actions with QuickPlay Pro.
5. Select one of the packets, taking note of the Source column, you can deduce whether this packet is from QuickPlay Pro, or from one of the nodes replying.
6. In the pane below, select `Data`, only this part is relevant for KiNet.
7. Search for patterns in the data and write them down.

Example for _Scan ethernet devices_ :

![image](https://github.com/Breina/ha-artnet-led/assets/3009914/c681e68f-e73a-468b-8934-cdfefb8dc959)

## Some notes on how the data will look

* It will most likely always start with `04 01 dc 4a`, this is a typical magic number and indicates it's KiNet.
* Some numbers take up more than one byte, there's two ways to store those. _Little endian_ stores it the small byte first, so `5` becomes `05 00`, while _big endian) stores the big byte first `00 05`. I think KiNet uses little endian. This example is for 2 bytes, might as well just be 1 or 4.
* The part after the magic number is the `01 00`, we think this is version, which in my case is V1. You should be on the lookout for `02 00`.
* In the part after that, we expect the _type_ of packet, also two bytes. This generally indicates the purpose of the packet and the further structure. In my example, `09 01`, is something we don't know yet, but it was the result of pushing the _scan_ button, so let's just call this specific packet like that. We are using `01 01` to send DMX data for V1. At a minimum, you should be looking out for what the V2 equivalent is for sending DMX data.
* The rest of my packet is static and I can't deduce any further meaning from `02 00 00 00 00 00 00 00`, so let's write this down into a format as such;

### KiNet Scan

```
Magic number            4 bytes                 `04 01 dc 4a`
Version                 2 bytes, little endian  `10 00` for V1
Type                    2 bytes                 `09 01`
Unknown			8 bytes			`02 00 00 00 00 00 00 00`
```

**Comment by Breina** [comment]
> I'm going to guess that contacting CK will lead to a dead end.

Just to follow up: you're right.

It requires signing an NDA, which makes it incompatible with open-source projects. Reverse engineering will be the only way, and even then they might strike it down.

**Comment by robbavel** [comment]
Can i contact you on discord by any chance? Or somewhere else private?


Reasons why i can discuss at that time
